### DISCUSSION TITLE
Question on how to update shopify application handle

### DISCUSSION THREAD

**Discussion by john-kuo** [kinngh/shopify-node-express-mongodb-app#122] [Q&A] [answered]
Hello @kinngh, I ran into another question on updating the application handle.
For example, given a shop admin URL at https://admin.shopify.com/store/skibummm/apps/app-handle.

Was looking at the app config [documentation](https://shopify.dev/docs/apps/tools/cli/configuration#global), it says we need to update it in the TOML file, but it looks like we are getting this config from the .env file.  I updated it inside .env and in shopify.js as 

`const shopify = shopifyApi({
  apiKey: process.env.SHOPIFY_API_KEY,
  apiSecretKey: process.env.SHOPIFY_API_SECRET,
  name: process.env.SHOPIFY_APP_NAME,
  handle: process.env.SHOPIFY_APP_HANDLE,
  scopes: process.env.SHOPIFY_API_SCOPES,
  hostName: process.env.SHOPIFY_APP_URL.replace(/https:\/\//, ""),
  hostScheme: "https",
  apiVersion: process.env.SHOPIFY_API_VERSION,
  isEmbeddedApp: true,
  logger: { level: isDev ? 1 : 0 }, //Error = 0,Warning = 1,Info = 2,Debug = 3
});`

I added the handle field, but it didn't update it correctly.

<img width="721" alt="Screenshot 2024-03-25 at 4 20 23 PM" src="https://github.com/kinngh/shopify-node-express-mongodb-app/assets/1480108/fcde5f05-a15b-49ac-83ec-906a10ff3556">

Any insights into this?  Thanks!

**Comment by kinngh** [comment]
You don't need to add handle to the `shopify.js` file at all. Just add in the handle to your `.env` as regular text like `my-app` or `next-project`. You might get the error that handle isn't unique and that means you'll have to change the handle, mostly adding a prefix works, and then you can run `npm run update:config` and it'll make the updates.

If your app already has a handle, add it as is. Going by your screenshot, your `.env` should look something like this:

```env
APP_NAME="Festival King"
APP_HANDLE="roulette-development"
```

**Reply by john-kuo** [reply]
ok, did you mean `npm run update:url`? there is no `npm run update:config` in the packagage.json file

**Reply by kinngh** [reply]
Totally forgot this repo doesn't have it yet - use the CLI to pull the config, update your handle and push it up^ Forget that you even need to do the whole `env` thing

**Reply by john-kuo** [reply]
I ran into this [error](https://community.shopify.com/c/shopify-apps/how-do-i-update-app-handle-slug-url/td-p/2456663) trying to update the handle field, updated the @shopify/cli-kit and @shopify/shopify-api to the latest..

There doesn't seem to be a solution atm

**Reply by kinngh** [reply]
First, run `shopify app dev --reset` and then link your project again, because the keys issue mostly relates to that. If that doesn't work, we're going to do some manual work with a script.

---
# TAKE A BACKUP OF YOUR CODE FIRST OR COPY PASTE THE FOLDER SO IF YOU MESS UP YOU CAN DISCARD CHANGES.
---

First install `@iarna/toml` as a dev dependency and then in your `_developer` make a new file called `tomlWriter.js` with this code:

```javascript
import toml from "@iarna/toml";
import "dotenv/config";
import fs from "fs";
import path from "path";

let config = {};

try {
  let appUrl = process.env.SHOPIFY_APP_URL;
  if (appUrl.endsWith("/")) {
    appUrl = appUrl.slice(0, -1);
  }
  // Globals
  config.name = process.env.APP_NAME;
  config.handle = process.env.APP_HANDLE;
  config.client_id = process.env.SHOPIFY_API_KEY;
  config.application_url = appUrl;
  config.embedded = true;
  config.extension_directories = ["extension/extensions/*"];

  // Auth
  config.auth = {};
  config.auth.redirect_urls = [
    `${appUrl}/api/auth/tokens`,
    `${appUrl}/api/auth/callback`,
  ];
  //Scopes
  config.access_scopes = {};
  config.access_scopes.scopes = process.env.SHOPIFY_API_SCOPES;
  config.access_scopes.use_legacy_install_flow = true;

  // Webhook event version to always match the app API version
  config.webhooks = {};
  config.webhooks.api_version = process.env.SHOPIFY_API_VERSION;

  // GDPR URLs
  config.webhooks.privacy_compliance = {};
  config.webhooks.privacy_compliance.customer_data_request_url = `${appUrl}/api/gdpr/customers_data_request`;
  config.webhooks.privacy_compliance.customer_deletion_url = `${appUrl}/api/gdpr/customers_redact`;
  config.webhooks.privacy_compliance.shop_deletion_url = `${appUrl}/api/gdpr/shop_redact`;

  // App Proxy
  if (
    process.env.APP_PROXY_PREFIX?.length > 1 &&
    process.env.APP_PROXY_SUBPATH?.length > 1
  ) {
    config.app_proxy = {};
    config.app_proxy.url = `${appUrl}/api/proxy_route`;
    config.app_proxy.prefix = process.env.APP_PROXY_PREFIX;
    config.app_proxy.subpath = process.env.APP_PROXY_SUBPATH;
  }

  // PoS
  if (process.env.POS_EMBEDDED?.length > 1) {
    config.pos = {};
    config.pos.embedded = process.env.POS_EMBEDDED === "true";
  }

  //Build
  config.build = {};
  config.build.include_config_on_deploy = true;

  //Write to toml
  let str = toml.stringify(config);
  str =
    "# Avoid writing to toml directly. Use your .env file instead\n\n" + str;

  fs.writeFileSync(path.join(process.cwd(), "shopify.app.toml"), str, (err) => {
    if (err) {
      console.log("An error occured while writing to file", e);
      return;
    }

    console.log("Written TOML to root");
    return;
  });

} catch (e) {
  console.error("---> An error occured while writing toml files");
  console.log(e.message);
}

```

Then in your `package.json` add a new script:
"update:config": "node _developer/tomlWriter.js && shopify app deploy",

Now before you run `update:config`, update your `.env` to this and fill it up:
```env
SHOPIFY_API_KEY=
SHOPIFY_API_SECRET=
SHOPIFY_API_SCOPES=
SHOPIFY_APP_URL=
SHOPIFY_API_VERSION=
DATABASE_URL=
ENCRYPTION_STRING=

# App Details
# App's name as in your Partner dashboard. Ex: "My App"
APP_NAME=
# App's URL that you want in the store. No spaces, use `-` instead. Ex: "my-app"
APP_HANDLE=

# App Proxy
APP_PROXY_PREFIX="apps"
APP_PROXY_SUBPATH=""
# Prefix can be `apps`, `a`, `community` or `tools`. Any other value will yield in errors.
# Proxy URL is autofilled
# If `APP_PROXY_PREFIX` or `APP_PROXY_SUBPATH` is left blank, no app proxy entry is created

POS_EMBEDDED=false

```

And now run `npm run update:config` and this should properly update the stuff.

**Reply by john-kuo** [reply]
thanks for the detailed instructions.  I tried both of them and they didn't work.

The `npm run update:config` still threw the same error...
<img width="841" alt="Screenshot 2024-03-25 at 7 20 21 PM" src="https://github.com/kinngh/shopify-node-express-mongodb-app/assets/1480108/24a0bae6-7b09-4f8a-a113-8ae78e9ebce9">

**Reply by kinngh** [reply]
What are you adding in as your handle?

**Reply by john-kuo** [reply]
APP_HANDLE=festival-king

**Reply by kinngh** [reply]
Even resetting doesn't work?

**Reply by kinngh** [reply]
Actually, show me your `toml` file too? Ensure you redact all the values except for handle

**Reply by john-kuo** [reply]
ya, resetting didn't work either

`
name = "Festival King"
handle = "festival-king"
client_id = ""
application_url = ""
embedded = true
extension_directories = [ "extension/extensions/*" ]

[auth]
redirect_urls = [
  "",
  ""
]

[access_scopes]
scopes = "write_products,read_themes,write_themes, read_discounts, write_discounts, read_price_rules, write_price_rules"
use_legacy_install_flow = true

[webhooks]
api_version = "2024-01"

  [webhooks.privacy_compliance]
  customer_data_request_url = ""
  customer_deletion_url = ""
  shop_deletion_url = ""

[app_proxy]
url = ""
prefix = "apps"
subpath = "express-proxy"

[pos]
embedded = false

[build]
include_config_on_deploy = true
`

**Reply by kinngh** [reply]
Can you DM me on [Twitter](https://twitter.com/kinngh) so we can setup a call? I want to see what can be done

**Reply by kinngh** [reply, marked as answer]
Answer: The user had the older `shopify-cli` installed from homebrew. Uninstalling the older gem and then installing the new CLI tools via `npm run g:install` script in `package.json` worked as expected.
